presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/joins/JoinGraph.java

```diff
@@ -0,0 +1,315 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.optimizations.joins;
+
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanNodeId;
+import com.facebook.presto.sql.planner.plan.PlanVisitor;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.tree.Expression;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.Multimap;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.sql.planner.optimizations.PruneIdentityProjections.isIdentityProjection;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * JoinGraph represents sequence of Joins, where nodes in the graph
+ * are PlanNodes that are being joined and edges are all equality join
+ * conditions between pair of nodes.
+ */
+public class JoinGraph
+{
+    private final Optional<Map<Symbol, Expression>> assignments;
+    private final List<Expression> filters;
+    private final List<PlanNode> nodes; // nodes in order of their appearance in plan
+    private final Multimap<PlanNode, Edge> edges;
```

Comment:
hmm, comparing references is good enough in this case, because `edges` are tightly related to `nodes`. But indeed using `PlanNodeId` might be less confusing.